engine/src/witness/common/block_height.rs

```diff
@@ -0,0 +1,86 @@
+use cf_chains::witness_period::SaturatingStep;
+use pallet_cf_elections::{
+	electoral_systems::block_height_witnesser::{
+		primitives::{Header, NonemptyContinuousHeaders},
+		ChainTypes, HeightWitnesserProperties,
+	},
+	ElectoralSystemTypes,
+};
+use sp_core::bounded::alloc::collections::VecDeque;
+
+#[async_trait::async_trait]
+pub trait HeaderClient<H> {
+	async fn best_block_header(&self) -> anyhow::Result<H>;
+	async fn block_header_by_height(&self, height: u64) -> anyhow::Result<H>;
+}
+
+pub async fn witness_headers<ES, C, RawHeader, Chain>(
+	client: &C,
+	properties: <ES as ElectoralSystemTypes>::ElectionProperties,
+	safety_buffer: u32,
+	header_conv: impl Fn(RawHeader) -> anyhow::Result<Header<Chain>>,
```

Comment:
I assume there was some reason of why we can't directly return a `Header<Chain>` from the HeaderClient? Seems like the `header_conv` should be part of the `HeaderClient` abstraction if possible...